### PR TITLE
[TR] PIM-4671: Fix family updater add identifier in requirement if not there

### PR DESCRIPTION
**Pull request by willy-ahva**
| Q                  | A
| ------------------ | ---
| Migration script?  | no
| Behats             | Blue
| Specs & Static     | OK
| Changelog updated  | no (should I ?)

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/MultipleOptionCodeUpdatedQueryGenerator.php [hunk 1]
```diff
@@ -37,6 +39,6 @@ public function generateQuery($entity, $field, $oldValue, $newValue)
     public function supports($entity, $field)
     {
         return parent::supports($entity, $field) &&
-            $entity->getAttribute()->getAttributeType() === 'pim_catalog_multiselect';
+            $entity->getAttribute()->getAttributeType() === AttributeTypes::OPTION_MULTI_SELECT;
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/MultipleOptionCodeUpdatedQueryGenerator.php]: yoda

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/MultipleOptionDeletedQueryGenerator.php [hunk 2]
```diff
@@ -37,6 +39,6 @@ public function generateQuery($entity, $field, $oldValue, $newValue)
     public function supports($entity, $field)
     {
         return parent::supports($entity, $field) &&
-            $entity->getAttribute()->getAttributeType() === 'pim_catalog_multiselect';
+            $entity->getAttribute()->getAttributeType() === AttributeTypes::OPTION_MULTI_SELECT;
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/MultipleOptionDeletedQueryGenerator.php]: yoda

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/MultipleOptionValueUpdatedQueryGenerator.php [hunk 3]
```diff
@@ -49,6 +51,6 @@ public function generateQuery($entity, $field, $oldValue, $newValue)
     public function supports($entity, $field)
     {
         return parent::supports($entity, $field) &&
-            $entity->getOption()->getAttribute()->getAttributeType() === 'pim_catalog_multiselect';
+            $entity->getOption()->getAttribute()->getAttributeType() === AttributeTypes::OPTION_MULTI_SELECT;
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/MultipleOptionValueUpdatedQueryGenerator.php]: yoda

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/OptionCodeUpdatedQueryGenerator.php [hunk 4]
```diff
@@ -37,6 +39,6 @@ public function generateQuery($entity, $field, $oldValue, $newValue)
     public function supports($entity, $field)
     {
         return parent::supports($entity, $field) &&
-            $entity->getAttribute()->getAttributeType() === 'pim_catalog_simpleselect';
+            $entity->getAttribute()->getAttributeType() === AttributeTypes::OPTION_SIMPLE_SELECT;
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/OptionCodeUpdatedQueryGenerator.php]: yoda

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/OptionDeletedQueryGenerator.php [hunk 5]
```diff
@@ -37,6 +39,6 @@ public function generateQuery($entity, $field, $oldValue, $newValue)
     public function supports($entity, $field)
     {
         return parent::supports($entity, $field) &&
-            $entity->getAttribute()->getAttributeType() === 'pim_catalog_simpleselect';
+            $entity->getAttribute()->getAttributeType() === AttributeTypes::OPTION_SIMPLE_SELECT;
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/OptionDeletedQueryGenerator.php]: yoda

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/OptionValueUpdatedQueryGenerator.php [hunk 6]
```diff
@@ -47,6 +49,6 @@ public function generateQuery($entity, $field, $oldValue, $newValue)
     public function supports($entity, $field)
     {
         return parent::supports($entity, $field) &&
-            $entity->getOption()->getAttribute()->getAttributeType() === 'pim_catalog_simpleselect';
+            $entity->getOption()->getAttribute()->getAttributeType() === AttributeTypes::OPTION_SIMPLE_SELECT;
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/OptionValueUpdatedQueryGenerator.php]: yoda

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Sorter/MetricSorter.php [hunk 7]
```diff
@@ -34,7 +35,7 @@ public function setQueryBuilder($queryBuilder)
      */
     public function supportsAttribute(AttributeInterface $attribute)
     {
-        return $attribute->getAttributeType() === 'pim_catalog_metric';
+        return $attribute->getAttributeType() === AttributeTypes::METRIC;
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Sorter/MetricSorter.php]: yoda

File: src/Pim/Bundle/CatalogBundle/Model/AbstractAttribute.php [hunk 8]
```diff
@@ -1002,7 +1003,7 @@ public function getReference()
     public function setAttributeType($type)
     {
         $this->attributeType = $type;
-        if ($this->attributeType === 'pim_catalog_identifier') {
+        if ($this->attributeType === AttributeTypes::IDENTIFIER) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Model/AbstractAttribute.php]: yoda

File: src/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/DateGuesser.php [hunk 9]
```diff
@@ -23,7 +24,7 @@ public function supportAttribute(AttributeInterface $attribute)
         return in_array(
             $attribute->getAttributeType(),
             array(
-                'pim_catalog_date',
+                AttributeTypes::DATE,
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/DateGuesser.php]: `AttributeTypes::DATE === $attribute->getAttributeType()` maybe ?

File: src/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/LengthGuesser.php [hunk 10]
```diff
@@ -29,9 +30,9 @@ public function supportAttribute(AttributeInterface $attribute)
         return in_array(
             $attribute->getAttributeType(),
             array(
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/LengthGuesser.php]: []

File: src/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RegexGuesser.php [hunk 11]
```diff
@@ -23,8 +24,8 @@ public function supportAttribute(AttributeInterface $attribute)
         return in_array(
             $attribute->getAttributeType(),
             array(
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Validator/ConstraintGuesser/RegexGuesser.php]: []

File: src/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidator.php [hunk 12]
```diff
@@ -35,7 +36,7 @@ public function __construct(AttributeRepositoryInterface $attributeRepository)
      */
     public function validate($attribute, Constraint $constraint)
     {
-        if ($attribute->getAttributeType() === 'pim_catalog_identifier') {
+        if ($attribute->getAttributeType() === AttributeTypes::IDENTIFIER) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidator.php]: yoda

File: src/Pim/Bundle/DataGridBundle/Datagrid/Product/ColumnsConfigurator.php [hunk 13]
```diff
@@ -137,7 +138,7 @@ protected function prepareAttributesColumns()
                     'groupOrder' => $attribute['groupOrder']
                 );
 
-                if ($attributeType === 'pim_catalog_identifier') {
+                if ($attributeType === AttributeTypes::IDENTIFIER) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/DataGridBundle/Datagrid/Product/ColumnsConfigurator.php]: yoda

File: src/Pim/Bundle/DataGridBundle/Datagrid/Product/FiltersConfigurator.php [hunk 14]
```diff
@@ -64,13 +65,13 @@ public function configure(DatagridConfiguration $configuration)
                 $filterConfig = $filterConfig + array(
                     ProductFilterUtility::DATA_NAME_KEY => $attributeCode,
                     'label'                             => $attribute['label'],
-                    'enabled'                           => ($attributeType === 'pim_catalog_identifier'),
+                    'enabled'                           => ($attributeType === AttributeTypes::IDENTIFIER),
```
Comment (phaseinducer) [inline src/Pim/Bundle/DataGridBundle/Datagrid/Product/FiltersConfigurator.php]: yoda

File: src/Pim/Bundle/DataGridBundle/Datagrid/Product/FiltersConfigurator.php [hunk 15]
```diff
@@ -64,13 +65,13 @@ public function configure(DatagridConfiguration $configuration)
...
                     'order'                             => $attribute['sortOrder'],
                     'group'                             => $attribute['group'],
                     'groupOrder'                        => $attribute['groupOrder']
                 );
 
-                if ($attributeType === 'pim_catalog_metric') {
+                if ($attributeType === AttributeTypes::METRIC) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/DataGridBundle/Datagrid/Product/FiltersConfigurator.php]: same

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/Product/OptionsTransformer.php [hunk 16]
```diff
@@ -25,7 +26,7 @@ public function transform(array $result, array $attribute, $locale, $scope)
     {
         $attributeCode = $attribute['code'];
         $normalizedData = $result['normalizedData'];
-        $fromNormData = array('pim_catalog_simpleselect', 'pim_catalog_multiselect');
+        $fromNormData = array(AttributeTypes::OPTION_SIMPLE_SELECT, AttributeTypes::OPTION_MULTI_SELECT);
```
Comment (phaseinducer) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/Product/OptionsTransformer.php]: []

File: src/Pim/Bundle/CatalogBundle/Model/AbstractProduct.php [hunk 17]
```diff
@@ -648,7 +646,7 @@ public function getMedia()
         foreach ($this->getValues() as $value) {
             if (in_array(
                 $value->getAttribute()->getAttributeType(),
-                array('pim_catalog_image', 'pim_catalog_file')
+                array(AttributeTypes::IMAGE, AttributeTypes::FILE)
```
Comment (grena) [inline src/Pim/Bundle/CatalogBundle/Model/AbstractProduct.php]: `[]`

File: src/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidator.php (hunk 12 above)
Comment (grena) [inline src/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidator.php]: yoda

File: src/Pim/Bundle/DataGridBundle/Datagrid/Product/ColumnsConfigurator.php (hunk 13 above)
Comment (grena) [inline src/Pim/Bundle/DataGridBundle/Datagrid/Product/ColumnsConfigurator.php]: yoda

File: src/Pim/Bundle/DataGridBundle/Datagrid/Product/FiltersConfigurator.php (hunk 14 above)
Comment (grena) [inline src/Pim/Bundle/DataGridBundle/Datagrid/Product/FiltersConfigurator.php]: yoda

File: src/Pim/Bundle/DataGridBundle/Datagrid/Product/FiltersConfigurator.php (hunk 15 above)
Comment (grena) [inline src/Pim/Bundle/DataGridBundle/Datagrid/Product/FiltersConfigurator.php]: yoda

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/Product/OptionsTransformer.php (hunk 16 above)
Comment (grena) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/Product/OptionsTransformer.php]: `[]`

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/Product/ValuesTransformer.php [hunk 18]
```diff
@@ -63,7 +64,7 @@ protected function prepareDateData(array $result, array $attribute)
         $backendType = $attribute['backendType'];
         $value = $result[$attributeCode];
 
-        if ($attribute['attributeType'] === 'pim_catalog_date' && isset($value[$backendType])) {
+        if ($attribute['attributeType'] === AttributeTypes::DATE && isset($value[$backendType])) {
```
Comment (grena) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/Product/ValuesTransformer.php]: yoda

File: src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/Product/ValuesTransformer.php [hunk 19]
```diff
@@ -82,7 +83,7 @@ protected function prepareMediaData(array $result, array $attribute)
         $attributeCode = $attribute['code'];
         $backendType = $attribute['backendType'];
         $value = $result[$attributeCode];
-        if ($attribute['attributeType'] === 'pim_catalog_image' && isset($value[$backendType])) {
+        if ($attribute['attributeType'] === AttributeTypes::IMAGE && isset($value[$backendType])) {
```
Comment (grena) [inline src/Pim/Bundle/DataGridBundle/Datasource/ResultRecord/MongoDbOdm/Product/ValuesTransformer.php]: yoda

File: src/Pim/Bundle/DataGridBundle/Extension/MassAction/Util/ProductFieldsBuilder.php [hunk 20]
```diff
@@ -145,9 +146,9 @@ protected function prepareAttributesList(array $attributesList)
                 }
             } elseif ($attribute->isScopable()) {
                 $fieldsList[] = sprintf('%s-%s', $attCode, $scopeCode);
-            } elseif ($attribute->getAttributeType() === 'pim_catalog_identifier') {
+            } elseif ($attribute->getAttributeType() === AttributeTypes::IDENTIFIER) {
```
Comment (grena) [inline src/Pim/Bundle/DataGridBundle/Extension/MassAction/Util/ProductFieldsBuilder.php]: yoda

File: src/Pim/Bundle/DataGridBundle/Extension/MassAction/Util/ProductFieldsBuilder.php [hunk 21]
```diff
@@ -145,9 +146,9 @@ protected function prepareAttributesList(array $attributesList)
...
                 array_unshift($fieldsList, $attCode);
-            } elseif ($attribute->getAttributeType() === 'pim_catalog_price_collection') {
+            } elseif ($attribute->getAttributeType() === AttributeTypes::PRICE_COLLECTION) {
```
Comment (grena) [inline src/Pim/Bundle/DataGridBundle/Extension/MassAction/Util/ProductFieldsBuilder.php]: yoda

File: src/Pim/Bundle/EnrichBundle/Controller/AttributeController.php [hunk 22]
```diff
@@ -354,7 +355,7 @@ protected function findAttributeOr404($id)
      */
     protected function validateRemoval(AttributeInterface $attribute)
     {
-        if ($attribute->getAttributeType() === 'pim_catalog_identifier') {
+        if ($attribute->getAttributeType() === AttributeTypes::IDENTIFIER) {
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Controller/AttributeController.php]: yoda

File: src/Pim/Bundle/EnrichBundle/Controller/FamilyController.php [hunk 23]
```diff
@@ -283,7 +284,7 @@ public function removeAttributeAction($familyId, $attributeId)
 
         if (false === $family->hasAttribute($attribute)) {
             throw new DeleteException($this->getTranslator()->trans('flash.family.attribute not found'));
-        } elseif ($attribute->getAttributeType() === 'pim_catalog_identifier') {
+        } elseif ($attribute->getAttributeType() === AttributeTypes::IDENTIFIER) {
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Controller/FamilyController.php]: yoda

File: src/Pim/Bundle/EnrichBundle/Form/Type/ConversionUnitsType.php [hunk 24]
```diff
@@ -45,7 +46,7 @@ public function buildForm(FormBuilderInterface $builder, array $options)
     {
         $metricAttributes = $this->entityManager
             ->getRepository($this->attributeClass)
-            ->findBy(array('attributeType' => 'pim_catalog_metric'));
+            ->findBy(array('attributeType' => AttributeTypes::METRIC));
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Form/Type/ConversionUnitsType.php]: `[]`

File: src/Pim/Bundle/PdfGeneratorBundle/Renderer/ProductPdfRenderer.php [hunk 25]
```diff
@@ -138,7 +138,7 @@ protected function getImageAttributes(ProductInterface $product, $locale)
         $attributes = [];
 
         foreach ($this->getAttributes($product, $locale) as $attribute) {
-            if ($attribute->getAttributeType() === static::IMAGE_ATTRIBUTE_TYPE) {
+            if ($attribute->getAttributeType() === AttributeTypes::IMAGE) {
```
Comment (grena) [inline src/Pim/Bundle/PdfGeneratorBundle/Renderer/ProductPdfRenderer.php]: yoda

File: src/Pim/Bundle/TransformBundle/Normalizer/Flat/ProductNormalizer.php [hunk 26]
```diff
@@ -126,7 +127,7 @@ protected function normalizeValues(ProductInterface $product, $format = null, ar
 
             foreach ($values as $value) {
                 $fieldValue = $this->getFieldValue($value);
-                if ($value->getAttribute()->getAttributeType() === 'pim_catalog_price_collection'
+                if ($value->getAttribute()->getAttributeType() === AttributeTypes::PRICE_COLLECTION
```
Comment (grena) [inline src/Pim/Bundle/TransformBundle/Normalizer/Flat/ProductNormalizer.php]: yoda

**Comment by nidup** [comment]
no need of upgrade changelog, it was a regression introduced during 1.4, GTM

**Pull request merged**